:sparkles: Add a "move to top" option to the context menu of the backlog

jwir3 opened the pull request:
By default, the backlog adds new stories to the bottom of the stories
listing. Current functionality does not include any way to move this story
to the top of the backlog, without dragging and dropping. This
functionality is suboptimal because, with many stories, it is often
laborious to move the story to the top of the backlog. Additionally, in
order to move into a sprint, it requires dragging a story to the part of
the backlog so that the sprints are not scrolled out of view.

This problem is resolved by this commit by providing a new context menu
option on stories in the backlog: move to top, which reorders the backlog
such that the selected story is moved to the top and all other stories
retain their relative ordering.

**Feature Example**:

![move-to-top](https://user-images.githubusercontent.com/997106/113491769-b787c400-9498-11eb-85bf-6a82b0d39074.gif)

Xaviju commented:
@jwir3 we are reviewing this PR with the team. Thanks!

Xaviju commented on package.json:
```diff
@@ -118,6 +117,7 @@
     "checksley": "git+https://github.com/jespino/checksley-bower.git",
     "dom-autoscroller": "^2.3.4",
     "dragula": "^3.7.2",
+    "Flot": "git+https://github.com/flot/flot.git#v0.8.3",
```
There is no change here, is it?

jwir3 commented:
There doesn't appear to be. I'm not sure why it made changes to the
package-lock and package.json. it's possible there were spacing changes or
something. I can remove them.

~Scott

On Tue, Apr 6, 2021, 6:37 AM Xaviju ***@***.***> wrote:

> *@Xaviju* commented on this pull request.
> ------------------------------
>
> In package.json
> <https://github.com/taigaio/taiga-front/pull/2239#discussion_r607771600>:
>
> > @@ -118,6 +117,7 @@
>      "checksley": "git+https://github.com/jespino/checksley-bower.git",
>      "dom-autoscroller": "^2.3.4",
>      "dragula": "^3.7.2",
> +    "Flot": "git+https://github.com/flot/flot.git#v0.8.3",
>
> There is no change here, is it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/taigaio/taiga-front/pull/2239#pullrequestreview-628867510>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHTN4SQD6C5YDJOPQMLJWDTHLXBDANCNFSM42KUQMHQ>
> .
>

Xaviju commented:
Thanks @jwir3 I'm gathering all the feedback, I'll try to send it later or tomorrow.

Xaviju commented:
Hi @jwir3 this is the feedback I've gathered so far. Would be nice if you can do this small fixes before merging:

- The 'Move to top' option in the dropdown should be the last option.
- It would be nice if you could add this SVG to the sprite and use them instead of the current arrow:
```svg
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16">
  <path d="M3.1 1a.7.7 0 000 1.4h9.3a.7.7 0 100-1.4H3zm4.7 3.2c-.2 0-.4 0-.5.2L2.5 9a.7.7 0 001 1l3.6-3.6v7.8c0 .4.3.7.7.7.3 0 .7-.3.7-.7V6.5l3.5 3.6a.7.7 0 001-1L8.2 4.4a.7.7 0 00-.4-.2z"/>
</svg>
```
- A small bug: if I move a User Story to the top, then move it somewhere else, cannot move it back to the top.
- Why do you reload the user stories? https://github.com/taigaio/taiga-front/pull/2239/files#diff-d3c6fc736cde8fa29ce4a4a674e7b716eb3ff496a183a5a9adc32e27c88c53f6R427

Thank you so much, this is a nice feature. :tada:

jwir3 commented:
>     * The 'Move to top' option in the dropdown should be the last option.

No problem. I'll make that change.

>     * It would be nice if you could add this SVG to the sprite and use them instead of the current arrow:

I can do this.

>     * A small bug: if I move a User Story to the top, then move it somewhere else, cannot move it back to the top.

Ah, I'll look into it. It's possible my logic is slightly incorrect. I'll check it out.

>     * Why do you reload the user stories? https://github.com/taigaio/taiga-front/pull/2239/files#diff-d3c6fc736cde8fa29ce4a4a674e7b716eb3ff496a183a5a9adc32e27c88c53f6R427

I thought this was the standard way, to be honest. I didn't realize that we don't need to reload the user stories after we make a change. I went through a couple of iterations, and, I _thought_ I needed to do this to make the ordering show up, but I could be wrong. I'll take a look at this and verify that it's actually necessary, and, if not, I'll remove it.

> Thank you so much, this is a nice feature. tada

Thanks! I'm on travel this week, but I will get this feedback addressed as soon as I can.

jwir3 commented:
@Xaviju One other thing I forgot to mention - I only added the localization for English (I'm only fluent in this language, unfortunately... I speak some French, but, it's... rusty... haha). I don't know what the procedure is for other localizations to be added. If you're ok with Google Translate, I can add _an_ entry for each of the localizations, but I doubt it will be very good.

jwir3 commented:
@Xaviju I fixed the issues you mentioned and removed the `package.json` and `package-lock.json` changes. However, I was not able to reproduce the issue you found where moving the user story to the top and then moving it somewhere else caused the move-to-top option to not be shown. Do you have a screencast you could show me? Perhaps I'm doing something different than you are. Here's an example of my workflow:

![move-to-top-bug](https://user-images.githubusercontent.com/997106/113898509-91f90400-9791-11eb-8b64-b2d0c0f4569e.gif)

jwir3 commented:
@Xaviju Ok, I think I fixed this issue.

What appears to have been happening is that the `first` class added to the sub-menu button in backlog rows was not being updated after it was set. (In other words, I think that when a backlog row is moved, it is rendered using the jade template, but subsequent rows don't seem to be re-rendered).

So, I added a code block that removes the `.first` from this button prior to moving the user story to the top. This _seems_ to have fixed the issue you're describing.

Let me know if you can think of a better solution... I'm not super familiar with jade templates and how the Taiga front infrastructure works, so it's possible there's another way to do this that I'm not seeing.

Xaviju commented:
Merged!
Thanks a lot for your contribution :tada:
